watertap/unit_models/zero_order/tests/test_pump_zo.py

```diff
@@ -0,0 +1,159 @@
+        data = model.db.get_unit_operation_parameters("pump")
+
+        model.fs.unit.load_parameters_from_database()
+
+        assert model.fs.unit.energy_electric_flow_vol_inlet.fixed
+        assert model.fs.unit.energy_electric_flow_vol_inlet.value == data[
+            "energy_electric_flow_vol_inlet"]["value"]
+
+    @pytest.mark.component
+    def test_degrees_of_freedom(self, model):
+        assert degrees_of_freedom(model.fs.unit) == 0
+
+    @pytest.mark.component
+    def test_unit_consistency(self, model):
+        assert_units_consistent(model.fs.unit)
+
+    @pytest.mark.component
+    def test_initialize(self, model):
+        initialization_tester(model)
+
+    @pytest.mark.solver
+    @pytest.mark.skipif(solver is None, reason="Solver not available")
+    @pytest.mark.component
+    def test_solution(self, model):
+        assert (pytest.approx(value(
+            model.fs.unit.inlet.flow_vol[0]), rel=1e-5) ==
+            value(model.fs.unit.outlet.flow_vol[0]))
+        for j in ["sulfur", "toc", "tss"]:
+            assert (pytest.approx(value(
+                model.fs.unit.inlet.conc_mass_comp[0, j]), rel=1e-5) ==
+                value(model.fs.unit.outlet.conc_mass_comp[0, j]))
+
+        assert (pytest.approx(10*0.102398663*3600, rel=1e-5) ==
+                value(model.fs.unit.electricity[0]))
+
+    @pytest.mark.component
+    def test_report(self, model, capsys):
+        model.fs.unit.report()
+
+        output = """
+====================================================================================
+Unit : fs.unit                                                             Time: 0.0
+------------------------------------------------------------------------------------
+    Unit Performance
+
+    Variables: 
+
+    Key                   : Value   : Fixed : Bounds
+       Electricity Demand :  3686.4 : False : (None, None)
+    Electricity Intensity : 0.10240 :  True : (None, None)
+
+------------------------------------------------------------------------------------
+    Stream Table
+                               Inlet  Outlet
+    Volumetric Flowrate         10      10  
+    Mass Concentration sulfur    1       1  
+    Mass Concentration toc       2       2  
+    Mass Concentration tss       3       3  
+====================================================================================
+"""
+
+        captured = capsys.readouterr()
+        assert output in captured.out
+
+
+db = Database()
+params = db._get_technology("pump")
+
+
+class TestPumpZOsubtype:
+    @pytest.fixture(scope="class")
+    def model(self):
+        m = ConcreteModel()
+
+        m.fs = FlowsheetBlock(default={"dynamic": False})
+        m.fs.params = WaterParameterBlock(
+            default={"solute_list": ["sulfur", "toc", "tss"]})
+
+        m.fs.unit = PumpZO(default={
+            "property_package": m.fs.params,
+            "database": db})
+
+        return m
+
+    @pytest.mark.parametrize("subtype", [params.keys()])
```

Comment:
I am sure this works just fine, but I don't understand how it works exactly. My understanding is that you somehow load in all pump subtypes from the yaml file, but I don't see the connection where that happens.

Reply:
There is some code at the module level above that gets the parameter dict, from which we create a list of top-level keys (the sub-types). This then tells `pytest` to run the test multiple times, each time using a different value from the list for the `subtype` argument in the line below.